### ISSUE TITLE
HamburgerMenu.IsFullScreen causing problems with behaviors

### ISSUE THREAD

**Issue by vivekhnz** [Windows-XAML/Template10#735]
I've been experiencing a number of strange quirks occurring within my app when changing the value of the `HamburgerMenu.IsFullScreen` property. Most notably, it appears that it can cause behaviors to stop working.

I think I've narrowed down the cause of this. In order to explain it better, I'll present it in a scenario similar to what I have in my app.
1. User opens app
2. `BootStrapper` navigates to splash screen
3. `BootStrapper` internally sets `IsFullScreen` to `true` because it's the splash screen
4. `BootStrapper` navigates to first page, `SignInPage`.
5. **SignInPage sets IsFullScreen to true** -> hamburger menu is hidden, everything appears fine
6. User signs in and app navigates to `HomePage`
7. **HomePage sets IsFullScreen to false** -> hamburger menu returns, but occasionally there are strange issues such as the Wide state being shown when at narrow width and vice versa
8. User navigates to `SettingsPage` and clicks the sign out button
9. App navigates back to `SignInPage` using `NavigationService.Navigate`
10. **SignInPage sets IsFullScreen to true** -> hamburger menu is hidden, but my behaviors aren't working! Also, some controls such as `ProgressRing` and `ProgressBar` aren't rendering.

So what happened? I think the issue lies in `UpdateFullScreen`.

``` c#
    private void UpdateFullScreen(bool? manual = null)
    {
        DebugWrite($"Mavnual: {manual}, IsFullScreen: {IsFullScreen}");

        var frame = NavigationService?.FrameFacade?.Frame;
        if (manual ?? IsFullScreen)
        {
            ShellSplitView.IsHitTestVisible = ShellSplitView.IsEnabled = false;
            ShellSplitView.Content = null;
            if (RootGrid.Children.Contains(ShellSplitView))
                RootGrid.Children.Remove(ShellSplitView);
            if (!RootGrid.Children.Contains(frame) && frame != null)
                RootGrid.Children.Add(frame);
        }
        else
        {
            ShellSplitView.IsHitTestVisible = ShellSplitView.IsEnabled = true;
            if (!RootGrid.Children.Contains(ShellSplitView))
                RootGrid.Children.Insert(0, ShellSplitView);
            if (RootGrid.Children.Contains(frame) && frame != null)
                RootGrid.Children.Remove(frame);
            ShellSplitView.Content = frame;
        }
    }
```

Here's what I think is causing this:
1. Changing the value of `IsFullScreen` fires the `UpdateFullScreen` method inside `HamburgerMenu.xaml.cs`.
   - The current value of `ShellSplitView.Content` is `NavigationService.FrameFacade.Frame`.
   - The current value of `NavigationService.FrameFacade.Frame.Content` is a `SignInPage`
   - _Therefore_, `ShellSplitView.Content.Content` is the `SignInPage`
2. `ShellSplitView.Content` is set to `null`, and the `SplitView` is removed from `RootGrid`.
   - The `SignInPage`'s parent was the frame
   - The frame's parent was the `SplitView`
   - The `SplitView` is no longer in the visual tree
   - _Therefore_, the **SignInPage is no longer in the visual tree**
3. **SignInPage.Unloaded is called**
   - The killing blow! `SignInPage.Unloaded` being called causes it's children to also be unloaded.
   - It also causes behaviors to be detached -> behaviors no longer work
4. The frame (which contains the `SignInPage`) is added directly to `RootGrid.Children`
   - `SignInPage` is back in the visual tree, as if nothing happened
   - `SignInPage.Loaded` is **NOT** called. This means the children elements are not `Loaded`, and the behaviors are not re-attached.

I think it's fairly plausible that this is the cause of a lot of recent issues other developers have been having with the `HamburgerMenu` control. Causing pages to think they're `Unloaded`, and then displaying them anyway seems like a likely candidate for strange behavior.

Hopefully this explanation was helpful. I'm not entirely sure of a solution to this issue. My best guess would be to eliminate the logic that replaces the content of `RootGrid` since that seems to be the culprit. Instead, it may be better to keep the `SplitView` around with the navigation frame as its content and just hide / disable the pane when `IsFullScreen` is changed.

**Comment by pabermod** [comment]
I have problems with HamburgerMenu.IsFullScreenl. I have a start page with the hamburger hidden, and in the OnNavigatedTo of the next page I show the Hamburger, but this is causing any progressring or progressbar to not show.

If I show the Hamburger in the OnNavigatedFrom of the first page it seems to work

**Comment by vivekhnz** [comment]
@kbronctjr I'm experiencing very similar behavior to this. I suspect it's due to the `Unloaded` event being called on the page, which then calls `Unloaded` on all child elements (including the `ProgressRing` and `ProgressBar`).

`ProgressRing` and `ProgressBar` probably have some sort of internal logic hooked to their `Unloaded` event that causes them to not render after being unloaded e.g. maybe the animation `Storyboard` is disposed.

I think the reason it works when you show the hamburger menu in `OnNavigatedFrom` is that the `NavigationService` may have not navigated yet and thus the frame content hasn't been set to the target page yet. As a result, the old page is `Unloaded`, not the new page.

While this works, I consider this solution somewhat inflexible. The hamburger menu's visibility should be defined by the page being displayed, not the page that we are navigating away from. What happens if the user can navigate from `SignInPage` to `HomePage` (which has a hamburger menu) and `AboutPage` (which doesn't have a hamburger menu)? You would need to store the target page within the view model somewhere in order to determine whether or not to show the hamburger menu.

You'd also need to replicate that same logic on every page. One way you could reduce this would be to have a `Dictionary<Type, bool>` somewhere that stores whether or not to show the hamburger menu for a target page. Then you perform a lookup every time you navigate and set the hamburger menu visibility accordingly. You could even put this logic in a subclass of `ViewModelBase` and make all of your page view models inherit from this subclass so you don't have to repeat the lookup everywhere.

Nonetheless, this seems like a fair amount of complexity for a workaround.

**Comment by JerryNixon** [comment]
I agree with your assessment of the problem. Any suggestions for a fix?

**Comment by vivekhnz** [comment]
I think the main problem here is that because the frame is being removed from the visual tree, it's being unloaded. I think some steps we could take would be:

- Go back to basics and remove the logic that replaces the content of `RootGrid` when `IsFullScreen` is changed
- Set `SplitView.Content` to `NavigationService..FrameFacade.Frame` when the `NavigationService` property is set (we already have this) BUT don't change it elsewhere
- Modify the `VisualState` setters to modify `HamburgerMenu.IsOpen` instead of `ShellSplitView.IsPaneOpen` and add additional logic to the `IsOpen
- When `IsFullScreen` is changed, modify the `Visibility` of `PaneContent`

**Comment by vivekhnz** [comment]
I think the main problem here is that because the frame is being removed from the visual tree, it's being unloaded. I think some steps we could take would be:
- Go back to basics and remove the logic that replaces the content of `RootGrid` when `IsFullScreen` is changed
- Set `SplitView.Content` to `NavigationService.FrameFacade.Frame` when the `NavigationService` property is set (we already have this) BUT don't change it elsewhere
- Modify the `VisualState` setters to modify `HamburgerMenu.IsOpen` instead of `ShellSplitView.IsPaneOpen` and add additional logic to the `IsOpen` setter to disallow the pane being opened if `IsFullScreen` is true
- Add similar logic for hiding the `HamburgerButton` (modify `HamburgerMenuVisibility` instead)
- When `IsFullScreen` is modified, send a property changed notification for `IsOpen` so it can adjust `ShellSplitView.IsPaneOpen`

I think that would be a good start. I haven't tested any of this yet. I could put together a pull request for this, but it probably wouldn't be until sometime next week. You're welcome to tackle it though and I'll try help out where I can :)

**Comment by angusbreno** [comment]
Hey guys, the first thing i wanna say is sorry. 

I could not wait the time of the PR's analysis and nuget generation so i forked this repo and i'm doing the changes there... That's why i set it up with continuous build and automated nuget publish with another name... 

https://github.com/NaylahProject/Template10.Reloaded

I don't know if @JerryNixon will accept those changes here. If u don't mind to take a look i will be glad to make the PR's here in the oficial/original repo.

There i fixed:
- FullScreen problems: 1.Changing fullscreen property with/without splash. 2.With/withou loading 3. Fixed hamburguer menu layout states without checking for loaded, using just the states triggers....
- With this changes above will be no more setters exceptions (that happened for the Splitview not be more in visualtree and visualstates tried to set values).
- Did some refrac on Hamburguer.
- https://github.com/Windows-XAML/Template10/issues/776
- https://github.com/Windows-XAML/Template10/issues/818
- https://github.com/Windows-XAML/Template10/issues/821
- https://github.com/Windows-XAML/Template10/issues/721

To test all the things and scenarios i've created a "little more complex app" with some fake-business logic,   all u need it's run PlaygoundApp to understand.

Thank you for this brilliant repo!

**Comment by schuettecarsten** [comment]
@angusbreno : Maybe you should create a PR for your changes? I've done some tests here, thank you for the idea with setting ShellSplitView.Visibility instead of removing it from the visual tree. This fixed some serious issues here, so I added a small PR for this.

**Comment by angusbreno** [comment]
Hey @carstenschuette i'm removing shellsplitview from visualtree, but i'm clearing the visualstates that is no longer needed. See at https://github.com/NaylahProject/Template10.Reloaded/blob/master/Template10%20(Library)/Controls/HamburgerMenu.xaml.cs#L690 pls.

**Comment by angusbreno** [comment]
I will see how i create a pull request with just some files change. Thank you @carstenschuette.

**Comment by JerryNixon** [comment]
Fixed in 1.1.10. If you see anything else, please open a new issue. Thanks.
